src/Documents/Session/DocumentSession.ts

```diff
@@ -109,43 +116,57 @@ export class DocumentSession implements IDocumentSession {
     .catch((error: RavenException) => PromiseResolver.reject(error, null, callback));
   }
 
-  public store(entity: IDocument, documentType?: IDocumentType, key?: DocumentKey, etag?: number, forceConcurrencyCheck?: boolean,
-     callback?: EntityCallback<IDocument>
-  ): Promise<IDocument> {
+  public store<T extends Object>(entity: T, key?: string, etag?: number, forceConcurrencyCheck?: boolean,
+     callback?: EntityCallback<T>
+  ): Promise<T> {
     this.incrementRequestsCount();
 
-    return this.prepareDocumentToStore(entity, documentType, key, etag, forceConcurrencyCheck)
-      .then((entity: IDocument) => {
+    return this.prepareDocumentToStore<T>(entity, key, etag, forceConcurrencyCheck)
+      .then((entity: T) => {
         let tag: number = null;
         const metadata = entity['@metadata'];
-        const conventions: DocumentConventions<IDocument> = this.conventions;
-        const documentKey: DocumentKey = conventions.tryGetIdFromInstance(entity);
+        const conventions: DocumentConventions = this.conventions;
+        const documentKey: string = conventions.tryGetIdFromInstance(entity);
 
         if (conventions.defaultUseOptimisticConcurrency || metadata['force_concurrency_check']) {
           tag = (metadata['@tag'] as number) || conventions.emptyEtag;
         }
 
         return this.requestsExecutor
           .execute(new PutDocumentCommand(documentKey, conventions.tryConvertToRawEntity(entity), tag))
-          .then((): IDocument => {
-            PromiseResolver.resolve<IDocument>(entity, null, callback);
-            return entity;
+          .then((): T => {
+            PromiseResolver.resolve<Object>(entity, null, callback);
+            return entity as T;
           });
       })
       .catch((error: RavenException) => PromiseResolver.reject(error, null, callback));
   }
 
-  public query(documentType?: IDocumentType, indexName?: string, usingDefaultOperator?: boolean, waitForNonStaleResults: boolean = false,
-     includes?: string[], withStatistics: boolean = false
-  ): IDocumentQuery {
-    return new DocumentQuery(this, this.requestsExecutor, documentType);
+  public query<T extends Object>(documentTypeOrObjectType?: string | DocumentConstructor<T>, indexName?: string, options?: IDocumentQueryOptions): IDocumentQuery<T> {
+    let usingDefaultOperator: QueryOperator = null;
+    let waitForNonStaleResults: boolean = null;
+    let includes: string[] = null;
+    let withStatistics: boolean = null;
+    let nestedObjectTypes: INestedObjectTypes = {} as INestedObjectTypes;
+
+    if (options) {
+      usingDefaultOperator = options.usingDefaultOperator || null;
+      waitForNonStaleResults = options.waitForNonStaleResults || null;
+      includes = options.includes || null;
+      withStatistics = options.withStatistics || null;
+      nestedObjectTypes = options.nestedObjectTypes || {};
+    }
+
+    return new DocumentQuery(this, this.requestsExecutor, documentTypeOrObjectType, indexName,
```

Comment:
it is optional, if you want - you can pass nestedObjectTypes hash which maps class constructor to nested object property name.
if you dont wont it - just dont pass it and all nested props will be plain Object

The same with documentTypeOrObjectType. if you pass a string 'Product' you will have plain object casted to Product. if you pass Product class constructor - you will have Product class instance